<h1>Import Libraries</h1>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

<h1>Load Dataset into DataFrame</h1>

In [2]:
DataFrame = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
DataFrame

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Check DataFrame Information</h1>

In [4]:
DataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h1>Basic Descriptive Statistic</h1>

In [5]:
DataFrame.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h1>Check and Fill Null Values</h1>

In [6]:
DataFrame.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
DataFrame['Latitude'] = DataFrame['Latitude'].fillna(DataFrame['Latitude'].mode()[0])
DataFrame['Longitude'] = DataFrame['Longitude'].fillna(DataFrame['Longitude'].mode()[0])
DataFrame['Road_Surface_Conditions'] = DataFrame['Road_Surface_Conditions'].fillna(DataFrame['Road_Surface_Conditions'].mode()[0])
DataFrame['Road_Type'] = DataFrame['Road_Type'].fillna(DataFrame['Road_Type'].mode()[0])
DataFrame['Urban_or_Rural_Area'] = DataFrame['Urban_or_Rural_Area'].fillna(DataFrame['Urban_or_Rural_Area'].mode()[0])
DataFrame['Weather_Conditions'] = DataFrame['Weather_Conditions'].fillna(DataFrame['Weather_Conditions'].mode()[0])

In [8]:
DataFrame.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [9]:
DataFrame.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [10]:
DataFrame['Accident Date'] = DataFrame['Accident Date'].astype('category')
DataFrame['Accident_Severity'] = DataFrame['Accident_Severity'].astype('category')
DataFrame['Light_Conditions'] = DataFrame['Light_Conditions'].astype('category')
DataFrame['District Area'] = DataFrame['District Area'].astype('category')
DataFrame['Road_Surface_Conditions'] = DataFrame['Road_Surface_Conditions'].astype('category')
DataFrame['Road_Type'] = DataFrame['Road_Type'].astype('category')
DataFrame['Urban_or_Rural_Area'] = DataFrame['Urban_or_Rural_Area'].astype('category')
DataFrame['Weather_Conditions'] = DataFrame['Weather_Conditions'].astype('category')


In [11]:
DataFrame.dtypes

Index                        object
Accident_Severity          category
Accident Date              category
Latitude                    float64
Light_Conditions           category
District Area              category
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type                 object
dtype: object

In [12]:
DataFrame['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [13]:
DataFrame['Accident Date'].value_counts()

Accident Date
30-11-2019    704
31-01-2019    697
13-11-2021    692
13-07-2019    692
14-08-2019    688
             ... 
30-12-2022    171
25-12-2019    157
25-12-2022    145
10/1/2022     123
25-12-2020    118
Name: count, Length: 1461, dtype: int64

In [14]:
DataFrame['Latitude'].value_counts()

Latitude
52.458798    75
52.949719    45
53.083165    33
52.967634    31
52.938860    29
             ..
56.880265     1
56.865367     1
57.528413     1
57.194499     1
51.493490     1
Name: count, Length: 511618, dtype: int64

In [15]:
DataFrame['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

In [16]:
DataFrame['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

In [17]:
DataFrame['Longitude'].value_counts()

Longitude
-0.977611    71
-1.871043    36
-0.816789    33
-1.190861    31
-1.216694    29
             ..
-4.169644     1
-3.798648     1
-3.630574     1
-3.836563     1
-3.924477     1
Name: count, Length: 529766, dtype: int64

In [18]:
DataFrame['Number_of_Casualties'].value_counts()

Number_of_Casualties
1     504041
2     107971
3      30776
4      11156
5       4123
6       1525
7        539
8        214
9        112
10        58
11        36
13        24
12        21
14        13
15        11
17         8
19         7
16         7
18         6
22         5
21         4
29         3
40         2
28         2
24         2
27         2
42         2
25         1
32         1
68         1
45         1
62         1
47         1
48         1
26         1
43         1
Name: count, dtype: int64

In [19]:
DataFrame['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
2     391995
1     200787
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
14         9
12         7
13         6
16         4
28         1
15         1
32         1
19         1
Name: count, dtype: int64

In [20]:
DataFrame['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     448547
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

In [21]:
DataFrame['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

In [22]:
DataFrame['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

In [23]:
DataFrame['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

In [24]:
DataFrame['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Clearing any Inconsistencies with the Data Set</h1>

In [25]:
DataFrame['Accident Date'] = DataFrame['Accident Date'].str.strip()
DataFrame['Accident Date'] = DataFrame['Accident Date'].astype('str')
DataFrame['Accident Date'] = DataFrame['Accident Date'].str.replace('/','-')

In [26]:
DataFrame['Accident Date'] = pd.to_datetime(DataFrame['Accident Date'], dayfirst = True, errors = 'coerce')

<h1>Extracting Date Information using Pandas Date Time</h1>

In [27]:
DataFrame['Year'] = DataFrame['Accident Date'].dt.year
DataFrame['Month'] = DataFrame['Accident Date'].dt.month
DataFrame['Day'] = DataFrame['Accident Date'].dt.day
DataFrame['DayOfWeek'] = DataFrame['Accident Date'].dt.dayofweek

In [28]:
DataFrame.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

In [29]:
DataFrame

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...  

<h1> Explanatory Data Analytics </h1>

<h3>1. Which District Area Have the Most Serious Accident and How many Fatal Accident happend in that Area?</h3>

In [30]:
DataFrame[DataFrame['Accident_Severity'] == 'Serious']['District Area'].value_counts()

District Area
Birmingham                   1474
Leeds                        1147
Westminster                   926
Bradford                      802
Sheffield                     749
                             ... 
Oswestry                       23
Orkney Islands                 19
Clackmannanshire               18
Shetland Islands               15
London Airport (Heathrow)       9
Name: count, Length: 422, dtype: int64

In [31]:
DataFrame[DataFrame['District Area'] == 'Birmingham']['Accident_Severity'].value_counts()

Accident_Severity
Slight     11912
Serious     1474
Fatal        105
Name: count, dtype: int64

<h3>Insight:</h3>
<h5>Birmingham have the most Serious Accident in all Districts and have 105 Fatal Accidents  </h4>
<hr>

<h3>2. Which Type of Vehicle has the least Slight Accident?</h3>

In [37]:
DataFrame[DataFrame['Accident_Severity'] == 'Slight']['Vehicle_Type'].min()

'Agricultural vehicle'

<h3>Insight:</h3>
<h5> The Agricultral Vehicles have the least of Slight Accident. </h4>
<hr>

<h3>3. </h3>

In [33]:
DataFrame.groupby(['Accident_Severity','Urban_or_Rural_Area','Road_Surface_Conditions'])['Light_Conditions'].count()

Accident_Severity  Urban_or_Rural_Area  Road_Surface_Conditions
Fatal              Rural                Dry                          3637
                                        Flood over 3cm. deep           20
                                        Frost or ice                  156
                                        Snow                           29
                                        Wet or damp                  1759
                   Unallocated          Dry                             0
                                        Flood over 3cm. deep            0
                                        Frost or ice                    0
                                        Snow                            0
                                        Wet or damp                     0
                   Urban                Dry                          2153
                                        Flood over 3cm. deep            3
                                        Frost or

<h3>Insight:</h3>
<h5></h4>
<hr>